### PR TITLE
Feat/1518 upload cert with new training

### PR DESCRIPTION
**Pull request by kapppa-joe**
#### Work done
- Amend frontend to allow adding new training together with certificate upload
- Add a measure to prevent submit button from being triggered more than once 

#### Tests
Does this PR include tests for the changes introduced?
- [x] Yes
- [ ] No, I found it difficult to test
- [ ] No, they are not required for this change

### CODE REVIEW

File: frontend/src/app/core/services/worker.service.ts [hunk 1]
```diff
@@ -234,7 +239,7 @@ export class WorkerService {
     trainingRecordId: string,
     record: TrainingRecordRequest,
   ) {
-    return this.http.put<TrainingRecordRequest>(
+    return this.http.put<any>(
```
Comment (duncanc19) [inline frontend/src/app/core/services/worker.service.ts:242]: Is the original interface still valid for this?

File: frontend/src/app/features/training-and-qualifications/add-edit-training/add-edit-training.component.spec.ts [hunk 2]
```diff
@@ -533,6 +555,85 @@ describe('AddEditTrainingComponent', () => {
...
+        userEvent.click(getByLabelText('Yes'));
+        const completedDate = getByTestId('completedDate');
+        userEvent.type(within(completedDate).getByLabelText('Day'), '10');
+        userEvent.type(within(completedDate).getByLabelText('Month'), '4');
+        userEvent.type(within(completedDate).getByLabelText('Year'), '2020');
+        const expiresDate = getByTestId('expiresDate');
+        userEvent.type(within(expiresDate).getByLabelText('Day'), '10');
+        userEvent.type(within(expiresDate).getByLabelText('Month'), '4');
+        userEvent.type(within(expiresDate).getByLabelText('Year'), '2027');
+        userEvent.type(getByLabelText('Notes'), 'Some notes added to this training');
```
Comment (duncanc19) [inline frontend/src/app/features/training-and-qualifications/add-edit-training/add-edit-training.component.spec.ts:588]: I'd be tempted to get rid of some of these inputs (maybe expiry and completed?) to make the test a bit shorter, as the important part is the two calls to the service being made

File: frontend/src/app/features/training-and-qualifications/add-edit-training/add-edit-training.component.ts [hunk 3]
```diff
@@ -189,8 +191,8 @@ export class AddEditTrainingComponent extends AddEditTrainingDirective implement
     if (this.trainingRecordId) {
       trainingRecordId = this.trainingRecordId;
     } else {
-      // TODO: this is the case of adding new training with certificate
-      // extract trainingRecordId from trainingRecordResponse
+      const { uid } = trainingRecordResponse as CreateTrainingRecordResponse;
+      trainingRecordId = uid;
```
Comment (duncanc19) [inline frontend/src/app/features/training-and-qualifications/add-edit-training/add-edit-training.component.ts:195]: I think you might be able to make this all a one liner by using an or but will need to confirm it works -  
`const trainingRecordId = this.trainingRecordId || trainingRecordResponse.uid`

File: frontend/src/app/shared/directives/add-edit-training/add-edit-training.directive.ts [hunk 4]
```diff
@@ -49,6 +49,7 @@ export class AddEditTrainingDirective implements OnInit, AfterViewInit {
   public showChangeLink: boolean = false;
   public multipleTrainingDetails: boolean;
   public trainingCertificates: TrainingCertificate[];
+  public disableSubmitButton: boolean = false;
```
Comment (duncanc19) [inline frontend/src/app/shared/directives/add-edit-training/add-edit-training.directive.ts:52]: Having the verb first maybe makes it sound like a function and a bit harder to tell what corresponds to true and false, could name it `submitButtonDisabled`?

File: frontend/src/app/core/services/worker.service.ts (hunk 1 above)
Comment (kapppa-joe) [reply frontend/src/app/core/services/worker.service.ts:242]: The original interface here probably wasn't quite right. The type in <> should denote the response object type, not the request object type. 
When I checked in the local apps, the response we get here seems to change depending on what fields are filled in, so I put this as <any> for the moment

**Review by duncanc19 — APPROVED** [review]

**Pull request merged**